### PR TITLE
chore: Add wrapper type for key strings

### PR DESCRIPTION
**Pull request by bruceg**
This change adds a newtype struct for the object key type called `KeyString` and a convenience type alias for the object mapping type called `ObjectMap`. This lays the groundwork for future changes to the key data model.

### CODE REVIEW

**Comment by pront** [comment]
I will review this today and hopefully we can get it merged before the next VRL release.

File: src/core/encode_key_value.rs [hunk 1]
```diff
@@ -154,13 +156,13 @@ impl fmt::Display for Data {
...
-    output: &'a mut BTreeMap<String, Data>,
+    output: &'a mut BTreeMap<KeyString, Data>,
 }
 
 impl<'a> KeyValueSerializer<'a> {
-    fn new(key: String, separator: char, output: &'a mut BTreeMap<String, Data>) -> Self {
+    fn new(key: KeyString, separator: char, output: &'a mut BTreeMap<KeyString, Data>) -> Self {
         Self {
-            key,
+            key: key.into(),
```
Comment (pront) [inline src/core/encode_key_value.rs:165]: Is it possible to implicitly convert here?

**Review by pront — APPROVED** [review]
LGTM. My only non blocking question is if we can go further and avoid using `.into()` everywhere.

File: src/core/encode_key_value.rs (hunk 1 above)
Comment (bruceg) [reply src/core/encode_key_value.rs]: Oh, good catch. The field in the struct should have been a `KeyString` as seen below in `fn process` this results in a double conversion. Fixed in f0c79084b

File: src/datadog/grok/parse_grok_rules.rs [hunk 2]
```diff
@@ -270,7 +270,7 @@ fn resolve_grok_pattern(
     }
 
     let match_name = &pattern.match_fn.name;
-    match context.aliases.get(match_name).cloned() {
+    match context.aliases.get(match_name.as_str()).cloned() {
```
[the remark this answers is not in the document]
Comment (bruceg) [reply src/datadog/grok/parse_grok_rules.rs:273]: We [already do](https://github.com/vectordotdev/vrl/pull/530/files/c2f0b4c6afb25df26928ac9dbdb58430e89f1f50#diff-a4c361e029b3a0812d7b832ab810fcd377851e901ffda8955d706e09e03d31fcR54), actually. I don't see how it helps here. `match_name` is a `String` and we need to borrow it as a `KeyString`. Rust doesn't know how to do that directly, so this needs a `.as_str()` to provide the intermediate between the two.

**Comment by bruceg** [comment]
> My only non blocking question is if we can go further and avoid using `.into()` everywhere.

Some of those are a feature, actually, to make explicit we are converting from some kind of string into a `KeyString`. Right now, that is just adding the wrapper type, but eventually we will be playing with the internals, making those `into`s necessary.

**Review by pront — APPROVED** [review]
Now that we published a new VRL release, we can get this merged.

**Pull request merged**